pageserver/src/remote_storage/rust_s3.rs

```diff
@@ -159,6 +159,31 @@ impl RemoteStorage for S3 {
         }
     }
 
+    async fn download_range(
+        &self,
+        from: &Self::StoragePath,
+        start_inclusive: u64,
+        end_inclusive: Option<u64>,
+        to: &mut (impl io::AsyncWrite + Unpin + Send + Sync),
+    ) -> anyhow::Result<()> {
+        let (data, code) = self
+            .bucket
+            .get_object_range(from.key(), start_inclusive, end_inclusive)
+            .await
+            .with_context(|| format!("Failed to download s3 object with key {}", from.key()))?;
+        if code != 200 {
+            Err(anyhow::format_err!(
+                "Received non-200 exit code during downloading object from directory, code: {}",
+                code
+            ))
+        } else {
+            io::copy(&mut io::BufReader::new(data.as_slice()), to)
+                .await
+                .context("Failed to write downloaded data into the destination buffer")?;
```

Comment:
Is streaming not allowed with get_object_range? Is that a limitation of our s3 crate?

If so, should we do `to.write_all`

Reply:
Alas, `rust_s3` returns `Vec<u8>` for both [`get_object`](https://github.com/durch/rust-s3/blob/c00a18693e53da096dd6e59d6e7880f2add6039f/s3/src/bucket.rs#L526) and [`get_object_range`](https://github.com/durch/rust-s3/blob/c00a18693e53da096dd6e59d6e7880f2add6039f/s3/src/bucket.rs#L608) 

I've used the `io::copy` version, keeping `rusoto` in mind, which has a streaming HTTP body returned instead, I'll place a comment about that and won't change the code itself, but let me know if you prefer writing vectors properly 🙂